In [1]:
from pandas.io.data import DataReader
# from pandas_datareader import data, wb
import pandas as pd
import numpy as np

/opt/conda/lib/python3.5/site-packages/pandas/io/data.py:35: FutureWarning: 
The pandas.io.data module is moved to a separate package (pandas-datareader) and will be removed from pandas in a future version.
After installing the pandas-datareader package (https://github.com/pydata/pandas-datareader), you can change the import ``from pandas.io import data, wb`` to ``from pandas_datareader import data, wb``.
  FutureWarning)


In [2]:
symbols = ['MSFT', 'GOOG', 'AAPL']

import datetime
start = datetime.datetime(2016, 12, 31)
end = datetime.datetime(2017, 1, 10)


## Data from Yahoo

In [3]:
data = dict((sym, DataReader(sym, "yahoo", start, end))
          for sym in symbols)

In [4]:
data['AAPL'].index


DatetimeIndex(['2017-01-03', '2017-01-04', '2017-01-05', '2017-01-06',
               '2017-01-09', '2017-01-10'],
              dtype='datetime64[ns]', name='Date', freq=None)

In [5]:
dfa = data['AAPL']
dfa['New']  = np.arange( dfa.shape[1] )

dfg = data['GOOG']
dfg['New'] = np.arange( dfg. shape[1] )[::-1]


# At the end of the day, Pandas are 2D
## even with Multi-Index
##    pct_change() works along row dimension (and is NOT affected by multi-index

In [6]:
c = pd.concat([ data['AAPL'], data['GOOG']], axis=0, keys=['AAPL', 'GOOG'])

In [7]:
c

Open        High         Low       Close    Volume  \
     Date                                                                   
AAPL 2017-01-03  115.800003  116.330002  114.760002  116.150002  28781900   
     2017-01-04  115.849998  116.510002  115.750000  116.019997  21118100   
     2017-01-05  115.919998  116.860001  115.809998  116.610001  22193600   
     2017-01-06  116.779999  118.160004  116.470001  117.910004  31751900   
     2017-01-09  117.949997  119.430000  117.940002  118.989998  33561900   
     2017-01-10  118.769997  119.379997  118.300003  119.110001  24462100   
GOOG 2017-01-03  778.809998  789.630005  775.799988  786.140015   1643100   
     2017-01-04  788.359985  791.340027  783.159973  786.900024   1065400   
     2017-01-05  786.080017  794.479980  785.020020  794.020020   1315400   
     2017-01-06  795.260010  807.900024  792.203979  806.150024   1620500   
     2017-01-09  806.400024  809.966003  802.830017  806.650024   1272400   
     2017-01-10  807.859985  809.130005  803.510010  804.789978   1174300   

                  Adj Close  New  
     Date                         
AAPL 2017-01-03  115.648597    0  
     2017-01-04  115.519154    1  
     2017-01-05  116.106611    2  
     2017-01-06  117.401002    3  
     2017-01-09  118.476334    4  
     2017-01-10  118.595819    5  
GOOG 2017-01-03  786.140015    5  
     2017-01-04  786.900024    4  
     2017-01-05  794.020020    3  
     2017-01-06  806.150024    2  
     2017-01-09  806.650024    1  
     2017-01-10  804.789978    0

## Notice below the bad value in GOOG Open on first date
## it is the change from the row above, which is AAPL last date

In [8]:
c.pct_change() * 100

Open        High         Low       Close     Volume  \
     Date                                                                    
AAPL 2017-01-03         NaN         NaN         NaN         NaN        NaN   
     2017-01-04    0.043174    0.154732    0.862668   -0.111929 -26.627151   
     2017-01-05    0.060423    0.300403    0.051834    0.508536   5.092788   
     2017-01-06    0.741892    1.112445    0.569902    1.114830  43.067821   
     2017-01-09    1.001882    1.074810    1.262128    0.915948   5.700446   
     2017-01-10    0.695210   -0.041868    0.305241    0.100851 -27.113483   
GOOG 2017-01-03  555.729576  561.442473  555.790337  560.011761 -93.283079   
     2017-01-04    1.226228    0.216560    0.948696    0.096676 -35.159150   
     2017-01-05   -0.289204    0.396789    0.237505    0.904816  23.465365   
     2017-01-06    1.167819    1.689161    0.915131    1.527670  23.194466   
     2017-01-09    1.400801    0.255722    1.341326    0.062023 -21.481024   
     2017-01-10    0.181047   -0.103214    0.084699   -0.230589  -7.709840   

                  Adj Close         New  
     Date                                
AAPL 2017-01-03         NaN         NaN  
     2017-01-04   -0.111928         inf  
     2017-01-05    0.508536  100.000000  
     2017-01-06    1.114830   50.000000  
     2017-01-09    0.915948   33.333333  
     2017-01-10    0.100851   25.000000  
GOOG 2017-01-03  562.873297    0.000000  
     2017-01-04    0.096676  -20.000000  
     2017-01-05    0.904816  -25.000000  
     2017-01-06    1.527670  -33.333333  
     2017-01-09    0.062023  -50.000000  
     2017-01-10   -0.230589 -100.000000

## Concat along columns

## The column is now multi-index

In [9]:
d=pd.concat( [data['AAPL'], data['GOOG']], axis=1, keys=['AAPL', 'GOOG'], names=["ticker", "attr"])

In [10]:
d

ticker            AAPL                                                \
attr              Open        High         Low       Close    Volume   
Date                                                                   
2017-01-03  115.800003  116.330002  114.760002  116.150002  28781900   
2017-01-04  115.849998  116.510002  115.750000  116.019997  21118100   
2017-01-05  115.919998  116.860001  115.809998  116.610001  22193600   
2017-01-06  116.779999  118.160004  116.470001  117.910004  31751900   
2017-01-09  117.949997  119.430000  117.940002  118.989998  33561900   
2017-01-10  118.769997  119.379997  118.300003  119.110001  24462100   

ticker                            GOOG                                      \
attr         Adj Close New        Open        High         Low       Close   
Date                                                                         
2017-01-03  115.648597   0  778.809998  789.630005  775.799988  786.140015   
2017-01-04  115.519154   1  788.359985  791.340027  783.159973  786.900024   
2017-01-05  116.106611   2  786.080017  794.479980  785.020020  794.020020   
2017-01-06  117.401002   3  795.260010  807.900024  792.203979  806.150024   
2017-01-09  118.476334   4  806.400024  809.966003  802.830017  806.650024   
2017-01-10  118.595819   5  807.859985  809.130005  803.510010  804.789978   

ticker                               
attr         Volume   Adj Close New  
Date                                 
2017-01-03  1643100  786.140015   5  
2017-01-04  1065400  786.900024   4  
2017-01-05  1315400  794.020020   3  
2017-01-06  1620500  806.150024   2  
2017-01-09  1272400  806.650024   1  
2017-01-10  1174300  804.789978   0

In [11]:
d.pct_change()

ticker          AAPL                                                    \
attr            Open      High       Low     Close    Volume Adj Close   
Date                                                                     
2017-01-03       NaN       NaN       NaN       NaN       NaN       NaN   
2017-01-04  0.000432  0.001547  0.008627 -0.001119 -0.266272 -0.001119   
2017-01-05  0.000604  0.003004  0.000518  0.005085  0.050928  0.005085   
2017-01-06  0.007419  0.011124  0.005699  0.011148  0.430678  0.011148   
2017-01-09  0.010019  0.010748  0.012621  0.009159  0.057004  0.009159   
2017-01-10  0.006952 -0.000419  0.003052  0.001009 -0.271135  0.001009   

ticker                    GOOG                                          \
attr             New      Open      High       Low     Close    Volume   
Date                                                                     
2017-01-03       NaN       NaN       NaN       NaN       NaN       NaN   
2017-01-04       inf  0.012262  0.002166  0.009487  0.000967 -0.351592   
2017-01-05  1.000000 -0.002892  0.003968  0.002375  0.009048  0.234654   
2017-01-06  0.500000  0.011678  0.016892  0.009151  0.015277  0.231945   
2017-01-09  0.333333  0.014008  0.002557  0.013413  0.000620 -0.214810   
2017-01-10  0.250000  0.001810 -0.001032  0.000847 -0.002306 -0.077098   

ticker                          
attr       Adj Close       New  
Date                            
2017-01-03       NaN       NaN  
2017-01-04  0.000967 -0.200000  
2017-01-05  0.009048 -0.250000  
2017-01-06  0.015277 -0.333333  
2017-01-09  0.000620 -0.500000  
2017-01-10 -0.002306 -1.000000

# To illustrate point that Pandas are 2D, try a different concat
# where Date is left-most part of multi-index rather than right-most

## move the ticker from the column multi-index to a row multi-index

In [12]:
e = d.stack(level=0)

In [13]:
e


attr                     Open        High         Low       Close    Volume  \
Date       ticker                                                             
2017-01-03 AAPL    115.800003  116.330002  114.760002  116.150002  28781900   
           GOOG    778.809998  789.630005  775.799988  786.140015   1643100   
2017-01-04 AAPL    115.849998  116.510002  115.750000  116.019997  21118100   
           GOOG    788.359985  791.340027  783.159973  786.900024   1065400   
2017-01-05 AAPL    115.919998  116.860001  115.809998  116.610001  22193600   
           GOOG    786.080017  794.479980  785.020020  794.020020   1315400   
2017-01-06 AAPL    116.779999  118.160004  116.470001  117.910004  31751900   
           GOOG    795.260010  807.900024  792.203979  806.150024   1620500   
2017-01-09 AAPL    117.949997  119.430000  117.940002  118.989998  33561900   
           GOOG    806.400024  809.966003  802.830017  806.650024   1272400   
2017-01-10 AAPL    118.769997  119.379997  118.300003  119.110001  24462100   
           GOOG    807.859985  809.130005  803.510010  804.789978   1174300   

attr                Adj Close  New  
Date       ticker                   
2017-01-03 AAPL    115.648597    0  
           GOOG    786.140015    5  
2017-01-04 AAPL    115.519154    1  
           GOOG    786.900024    4  
2017-01-05 AAPL    116.106611    2  
           GOOG    794.020020    3  
2017-01-06 AAPL    117.401002    3  
           GOOG    806.150024    2  
2017-01-09 AAPL    118.476334    4  
           GOOG    806.650024    1  
2017-01-10 AAPL    118.595819    5  
           GOOG    804.789978    0

## NOTICE below: the bad pct change for GOOG Open on first day
### This is the pct change from AAPL first day Open (115.8) to GOOG first day Open (778.8)
###   This is b/c they are consecutive rows (i.e., multi-index is just a fancy way of labelling the row, it has not import for computation)

In [14]:
e.pct_change() * 100

attr                     Open        High         Low       Close  \
Date       ticker                                                   
2017-01-03 AAPL           NaN         NaN         NaN         NaN   
           GOOG    572.547477  578.784485  576.019497  576.831684   
2017-01-04 AAPL    -85.124742  -85.244988  -85.079917  -85.241815   
           GOOG    580.500646  579.203513  576.596089  578.245168   
2017-01-05 AAPL    -85.296058  -85.232644  -85.212472  -85.181091   
           GOOG    578.122870  579.856215  577.851683  580.919315   
2017-01-06 AAPL    -85.144006  -85.127378  -85.163436  -85.150248   
           GOOG    580.989910  583.733917  580.178563  583.699429   
2017-01-09 AAPL    -85.168373  -85.217230  -85.112420  -85.239720   
           GOOG    583.679563  578.193086  580.710534  577.914142   
2017-01-10 AAPL    -85.271578  -85.261110  -85.264626  -85.233993   
           GOOG    580.188604  577.776868  579.213854  575.669525   

attr                    Volume   Adj Close         New  
Date       ticker                                       
2017-01-03 AAPL            NaN         NaN         NaN  
           GOOG     -94.291204  579.766150         inf  
2017-01-04 AAPL    1185.259570  -85.305524  -80.000000  
           GOOG     -94.955039  581.185757  300.000000  
2017-01-05 AAPL    1983.123709  -85.245062  -50.000000  
           GOOG     -94.073066  583.871498   50.000000  
2017-01-06 AAPL    2313.858902  -85.214352    0.000000  
           GOOG     -94.896368  586.663666  -33.333333  
2017-01-09 AAPL    1971.082999  -85.303439  100.000000  
           GOOG     -96.208796  580.853295  -75.000000  
2017-01-10 AAPL    1822.516504  -85.297736  400.000000  
           GOOG     -95.199513  578.598946 -100.000000

## The below won't work either for pct_chg() b/c row order not consistent with what we want

In [15]:
f = e.swaplevel(i=0,j=1)

In [16]:
f

attr                     Open        High         Low       Close    Volume  \
ticker Date                                                                   
AAPL   2017-01-03  115.800003  116.330002  114.760002  116.150002  28781900   
GOOG   2017-01-03  778.809998  789.630005  775.799988  786.140015   1643100   
AAPL   2017-01-04  115.849998  116.510002  115.750000  116.019997  21118100   
GOOG   2017-01-04  788.359985  791.340027  783.159973  786.900024   1065400   
AAPL   2017-01-05  115.919998  116.860001  115.809998  116.610001  22193600   
GOOG   2017-01-05  786.080017  794.479980  785.020020  794.020020   1315400   
AAPL   2017-01-06  116.779999  118.160004  116.470001  117.910004  31751900   
GOOG   2017-01-06  795.260010  807.900024  792.203979  806.150024   1620500   
AAPL   2017-01-09  117.949997  119.430000  117.940002  118.989998  33561900   
GOOG   2017-01-09  806.400024  809.966003  802.830017  806.650024   1272400   
AAPL   2017-01-10  118.769997  119.379997  118.300003  119.110001  24462100   
GOOG   2017-01-10  807.859985  809.130005  803.510010  804.789978   1174300   

attr                Adj Close  New  
ticker Date                         
AAPL   2017-01-03  115.648597    0  
GOOG   2017-01-03  786.140015    5  
AAPL   2017-01-04  115.519154    1  
GOOG   2017-01-04  786.900024    4  
AAPL   2017-01-05  116.106611    2  
GOOG   2017-01-05  794.020020    3  
AAPL   2017-01-06  117.401002    3  
GOOG   2017-01-06  806.150024    2  
AAPL   2017-01-09  118.476334    4  
GOOG   2017-01-09  806.650024    1  
AAPL   2017-01-10  118.595819    5  
GOOG   2017-01-10  804.789978    0

In [42]:
bydate = c.groupby(level=1)
for name, group in bydate:
        print("Group name: {gname}, group:\n\t{ggroup}\n\n".format(gname=name, ggroup=group))

Group name: 2017-01-03 00:00:00, group:
	                       Open        High         Low       Close    Volume  \
     Date                                                                   
AAPL 2017-01-03  115.800003  116.330002  114.760002  116.150002  28781900   
GOOG 2017-01-03  778.809998  789.630005  775.799988  786.140015   1643100   

                  Adj Close  New  
     Date                         
AAPL 2017-01-03  115.648597    0  
GOOG 2017-01-03  786.140015    5  


Group name: 2017-01-04 00:00:00, group:
	                       Open        High         Low       Close    Volume  \
     Date                                                                   
AAPL 2017-01-04  115.849998  116.510002  115.750000  116.019997  21118100   
GOOG 2017-01-04  788.359985  791.340027  783.159973  786.900024   1065400   

                  Adj Close  New  
     Date                         
AAPL 2017-01-04  115.519154    1  
GOOG 2017-01-04  786.900024    4  


Group name: 2017

In [40]:
byticker = c.groupby(level=0)
for name, group in byticker:
        print("Group name: {gname}, group:\n\t{ggroup}\n\n".format(gname=name, ggroup=group))

Group name: AAPL, group:
	                       Open        High         Low       Close    Volume  \
     Date                                                                   
AAPL 2017-01-03  115.800003  116.330002  114.760002  116.150002  28781900   
     2017-01-04  115.849998  116.510002  115.750000  116.019997  21118100   
     2017-01-05  115.919998  116.860001  115.809998  116.610001  22193600   
     2017-01-06  116.779999  118.160004  116.470001  117.910004  31751900   
     2017-01-09  117.949997  119.430000  117.940002  118.989998  33561900   
     2017-01-10  118.769997  119.379997  118.300003  119.110001  24462100   

                  Adj Close  New  
     Date                         
AAPL 2017-01-03  115.648597    0  
     2017-01-04  115.519154    1  
     2017-01-05  116.106611    2  
     2017-01-06  117.401002    3  
     2017-01-09  118.476334    4  
     2017-01-10  118.595819    5  


Group name: GOOG, group:
	                       Open        High         L